server/controllers/orderControllers.js

```diff
@@ -0,0 +1,154 @@
+    const insufficientStockProducts = [];
+
+    // Check stock for all products
+    for (let item of products) {
+      const product = await Product.findById(item.productId);
+      if (!product) {
+        return res.status(404).send(`Product not found: ${item.productId}`);
+      }
+
+      if (product.stock < item.quantity) {
+        insufficientStockProducts.push(product.title); // Collect product titles with insufficient stock
+      }
+    }
+
+    // If there are any products with insufficient stock, return an error message
+    if (insufficientStockProducts.length > 0) {
+      return res
+        .status(400)
+        .send(
+          `Not enough stock for product(s): ${insufficientStockProducts.join(
+            ", "
+          )}`
+        );
+    }
+
+    // Create the order if all products have sufficient stock
+    const newOrder = new Order({
+      userId,
+      products,
+      amount,
+      address,
+    });
+
+    const order = await newOrder.save();
+
+    // Deduct stock for all products after creating the order
+    for (let item of products) {
+      const product = await Product.findById(item.productId);
+      product.stock -= item.quantity;
+      await product.save();
+    }
+
+    res.status(201).json(order);
+  } catch (error) {
+    console.error("Error inserting document into collection:", error);
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
+
+export const updateOrder = async (req, res) => {
+  const orderId = req.params.id;
+  const { userId, products, totalAmount, status } = req.body;
+
+  try {
+    // Find the order by ID
+    const order = await Order.findById(orderId);
+    if (!order) {
+      return res.status(404).send("Order not found");
+    }
+
+    if (products) {
+      // First, restore stock levels for the products in the current order
+      for (let item of order.products) {
+        const product = await Product.findById(item.productId);
+        if (product) {
+          product.stock += item.quantity;
+          await product.save();
+        }
+      }
+
+      // Check new product stock levels
+      for (let item of products) {
+        const product = await Product.findById(item.productId);
+        if (!product) {
+          return res.status(404).send(`Product not found: ${item.productId}`);
+        }
+        if (product.stock < item.quantity) {
+          return res
+            .status(400)
+            .send(`Not enough stock for product: ${product.name}`);
+        }
+      }
+
+      // Deduct stock levels for the new products
+      for (let item of products) {
+        const product = await Product.findById(item.productId);
+        product.stock -= item.quantity;
+        await product.save();
+      }
+
+      // Update the order's products
+      order.products = products;
+    }
+
+    // Update other fields if provided
+    if (userId) order.userId = userId;
+    if (totalAmount) order.amount = totalAmount;
```

Comment:
your order payload doens't show `totalAmount` anywhere.

Reply:
it has been ratified